ipa-core/src/protocol/hybrid/agg.rs

```diff
@@ -0,0 +1,217 @@
+use std::collections::HashMap;
+
+use crate::{
+    error::Error,
+    ff::{boolean::Boolean, boolean_array::BooleanArray, ArrayAccess},
+    helpers::TotalRecords,
+    protocol::{
+        boolean::step::{EightBitStep, ThirtyTwoBitStep},
+        context::{
+            dzkp_validator::{DZKPValidator, TARGET_PROOF_SIZE},
+            Context, DZKPUpgraded, MaliciousProtocolSteps, UpgradableContext,
+        },
+        hybrid::step::HybridStep,
+        ipa_prf::boolean_ops::addition_sequential::integer_add,
+        BooleanProtocols, RecordId,
+    },
+    report::hybrid::{AggregateableHybridReport, PrfHybridReport},
+    secret_sharing::{replicated::semi_honest::AdditiveShare as Replicated, SharedValue},
+};
+
+pub async fn aggregate_reports<BK, V, C>(
+    ctx: C,
+    reports: Vec<PrfHybridReport<BK, V>>,
+) -> Result<Vec<AggregateableHybridReport<BK, V>>, Error>
+where
+    C: UpgradableContext,
+    BK: SharedValue + BooleanArray,
+    V: SharedValue + BooleanArray,
+    Replicated<Boolean>: BooleanProtocols<DZKPUpgraded<C>>,
+{
+    let mut reports_by_matchkey = HashMap::new();
+    reports_by_matchkey.reserve(reports.len() / 2);
+
+    for report in reports {
+        reports_by_matchkey
+            .entry(report.match_key)
+            .or_insert_with(Vec::new)
+            .push(report);
+    }
+
+    // an honest client and report collector will only submit
+    // one report with a breakdown key and one report with a value.
+    // if there are less, it's unattributed. if more, something went wrong.
+    // we remove these (instead of erroring/panicing).
+    let report_pairs: Vec<[AggregateableHybridReport<BK, V>; 2]> = reports_by_matchkey
```

Comment:
it should be possible to filter out reports as you are adding them to the hashmap. That eliminates the need for a vector inside that map and postprocessing. @cberkhoff had somewhat similar problem to tackle in the aggregation, so that code can potentially be reused here

Reply:
Another way to do this would be to have the values in the hashmap be a tuple: `([AggregateableHybridReport<BK, V>; 2], usize)`. where the second value is a count of the number of times that `match_key` appears. then we grab a vector of the values from that hashmap and filter down to instances where the count is exactly 2. 

that should be a bit more memory efficient since we'll have an upper bound on the size and know that every value is exactly an array with two reports and a usize.

Reply:
you may run into Rust issues here - the most straightforward way to go around would be to use `[Option<Report>; 2]` as the value in that hashmap. It would be a bit clunky to deal with this array having 1 or 2 items, so you could create a wrapper type for it

Reply:
take a look at this, it uses an fixed array and count, and then filters the rows without count == 2 at the end.

Reply:
it works but it would be better if this internal state magic happens somewhere outside of this function. Something like breakdown reveal aggregation struct would be great to have here